### PR TITLE
Support Angstrom units with Molecule class and molecular_hamiltonian function

### PR DESCRIPTION
**Pull request by ddhawan11**
**Context:**
Support Angstrom units along with Bohr

**Description of the Change:**
Molecule class and molecular_hamiltonian function now works with both Bohr and Angstrom units.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by ddhawan11** [comment]
[sc-62838]

**Review by austingmhuang — COMMENTED** [review]
Looks good @ddhawan11.

Maybe I missed it (or it already exists) but is there some kind of warning or error that will pop up when a user inputs a unit that doesn't exist?

For example, let's say they put a typo and wrote angstorm instead of angstrom.
`molecule_ang = qchem.molecular_hamiltonian(symbols, geometry_ang, unit="Angstorm")`.

Would this raise a warning and let them know 'Bohr' was used instead?

File: pennylane/qchem/openfermion_obs.py [hunk 1]
```diff
@@ -1010,6 +1012,9 @@ def _(
     load_data=False,
     convert_tol=1e012,
 ):
+    if unit == "Angstrom":
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:1015]: We should add an `if` condition for inconsistent unit values.

File: pennylane/qchem/molecule.py [hunk 2]
```diff
@@ -121,7 +121,7 @@ def __init__(
 
         if unit == "Angstrom":
             self.coordinates = self.coordinates / bohr_angs
-        elif unit != "Angstrom" and unit != "Bohr":
+        elif unit not in ("Angstrom", "Bohr"):
```
Comment (obliviateandsurrender) [inline pennylane/qchem/molecule.py:124]: We should always aim to fail early, i.e., we should shift this right under the checks done above for atoms and basis set.

File: pennylane/qchem/molecule.py [hunk 3]
```diff
@@ -97,12 +102,21 @@ def __init__(
...
         self.n_basis, self.basis_data = mol_basis_data(self.basis_name, self.symbols, load_data)
 
+        if self.unit == "Angstrom":
+            self.coordinates = self.coordinates / bohr_angs
+        elif self.unit not in ("Angstrom", "Bohr"):
+            raise ValueError(
+                f"The provided unit, '{unit}' is not supported. "
+                f"Please set 'unit' to 'Bohr' or 'Angstrom'."
+            )
+
```
Comment (austingmhuang) [inline pennylane/qchem/molecule.py:119]: The string comparison here is case-sensitive. We should make this case insensitive 😄

File: pennylane/qchem/molecule.py [hunk 4]
```diff
@@ -97,12 +102,21 @@ def __init__(
...
         self.name = name
         self.n_basis, self.basis_data = mol_basis_data(self.basis_name, self.symbols, load_data)
 
+        if self.unit == "Angstrom":
+            self.coordinates = self.coordinates / bohr_angs
+        elif self.unit not in ("Angstrom", "Bohr"):
+            raise ValueError(
+                f"The provided unit, '{unit}' is not supported. "
+                f"Please set 'unit' to 'Bohr' or 'Angstrom'."
+            )
```
Comment (austingmhuang) [inline pennylane/qchem/molecule.py:118]:
```suggestion
        if self.unit not in ("angstrom", "bohr"):
            raise ValueError(
                f"The provided unit, '{unit}' is not supported. "
                f"Please set 'unit' to 'Bohr' or 'Angstrom'."
            )
            
        if self.unit == "angstrom":
            self.coordinates = self.coordinates / bohr_angs
```

We should raise errors early.

File: pennylane/qchem/openfermion_obs.py [hunk 5]
```diff
@@ -1013,6 +1014,15 @@ def _(
...
 ):
+
+    if unit == "Angstrom":
+        coordinates = coordinates / bohr_angs
+    elif unit not in ("Angstrom", "Bohr"):
+        raise ValueError(
+            f"The provided unit, '{unit}' is not supported. "
+            f"Please set 'unit' to 'Bohr' or 'Angstrom'."
+        )
+
```
Comment (austingmhuang) [inline pennylane/qchem/openfermion_obs.py:1025]:
```suggestion
    if unit not in ("angstrom", "bohr"):
        raise ValueError(
            f"The provided unit, '{unit}' is not supported. "
            f"Please set 'unit' to 'Bohr' or 'Angstrom'."
        )
        
    if unit == "angstrom":
        coordinates = coordinates / bohr_angs

```

Should make this not case sensitive either.

**Review by austingmhuang — COMMENTED** [review]
Looks great @ddhawan11! Just some small changes but otherwise should be good to go soon.

File: pennylane/qchem/openfermion_obs.py [hunk 6]
```diff
@@ -1013,6 +1014,16 @@ def _(
     load_data=False,
     convert_tol=1e12,
 ):
+
+    if unit not in ("angstrom", "bohr"):
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:1018]:
```suggestion
    if unit.lower() not in ("angstrom", "bohr"):
```

**Review by austingmhuang — APPROVED** [review]
Looks good to me! Nice work @ddhawan11 :)

File: pennylane/qchem/openfermion_obs.py [hunk 7]
```diff
@@ -1013,6 +1014,16 @@ def _(
...
 ):
+
+    if unit.strip().lower() not in ("angstrom", "bohr"):
+        raise ValueError(
+            f"The provided unit, '{unit}' is not supported. "
+            f"Please set 'unit' to 'Bohr' or 'Angstrom'."
+        )
+
+    if unit.strip().lower() == "angstrom":
+        coordinates = coordinates / bohr_angs
```
Comment (obliviateandsurrender) [inline pennylane/qchem/openfermion_obs.py:1025]:
```suggestion
    if (coord_unit := unit.strip().lower()) not in ("angstrom", "bohr"):
        raise ValueError(
            f"The provided unit, '{unit}' is not supported. "
            f"Please set 'unit' to 'Bohr' or 'Angstrom'."
        )

    if coord_unit == "angstrom":
        coordinates = coordinates / bohr_angs
```
We might want to repeat the stripping.

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 8]
```diff
@@ -782,3 +782,83 @@ def test_error_raised_for_missing_molecule_information():
...
+    )
+
+    hamiltonian_ang, _ = qchem.molecular_hamiltonian(
+        symbols,
+        geometry_ang,
+        unit="Angstrom",
+        method=method,
+        outpath=tmpdir.strpath,
+    )
+    assert qml.ops.functions.equal(hamiltonian_ang, hamiltonian_bohr)
```
Comment (obliviateandsurrender) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:820]: Could we not use `qml.equal` directly here, or is this some other function? 🤔

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 9]
```diff
@@ -782,3 +782,83 @@ def test_error_raised_for_missing_molecule_information():
...
+)
+def test_coordinate_units_for_molecular_hamiltonian_molecule_class(method, tmpdir):
+    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units
+    based on user defined information.
+    """
+
+    symbols = ["H", "H"]
+    geometry_bohr = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+    geometry_ang = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.529177210903]])
+    method = "pyscf"
```
Comment (obliviateandsurrender) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:839]:
```suggestion
```
Isn't this overriding the parameterized `method`?

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 10]
```diff
@@ -782,3 +782,83 @@ def test_error_raised_for_missing_molecule_information():
...
+        outpath=tmpdir.strpath,
+    )
+
+    molecule_ang = qchem.Molecule(symbols, geometry_ang, unit="Angstrom")
+    hamiltonian_ang, _ = qchem.molecular_hamiltonian(
+        molecule_ang,
+        method=method,
+        outpath=tmpdir.strpath,
+    )
+    assert qml.ops.functions.equal(hamiltonian_ang, hamiltonian_bohr)
```
Comment (obliviateandsurrender) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:854]: Same as above.

**Review by obliviateandsurrender — CHANGES_REQUESTED** [review]
Just a couple of more changes, and should be good to go! 🚀

File: doc/releases/changelog-dev.md [hunk 11]
```diff
@@ -71,6 +71,9 @@
   `qml.devices.Device`, which follows the new device API.
   [(#5581)](https://github.com/PennyLaneAI/pennylane/pull/5581)
 
+* `qml.qchem.Molecule` and qchem functions that take Molecule object as an argument now work with coordinates
```
[the remark this answers is not in the document]
Comment (ddhawan11) [reply doc/releases/changelog-dev.md:74]: This is just a placeholder for now, we will be creating one change-log entry for all of the qchem feature branch.

**Review by obliviateandsurrender — APPROVED** [review]
Good work @ddhawan11, LGTM!

File: pennylane/qchem/molecule.py [hunk 12]
```diff
@@ -75,6 +79,7 @@ def __init__(
         alpha=None,
         coeff=None,
         normalize=True,
+        unit="Bohr",
```
Comment (soranjh) [inline pennylane/qchem/molecule.py:82]:
```suggestion
        unit="bohr",
```

File: pennylane/qchem/molecule.py [hunk 13]
```diff
@@ -97,12 +102,22 @@ def __init__(
...
         self.charge = charge
         self.mult = mult
         self.basis_name = basis_name.lower()
         self.name = name
         self.n_basis, self.basis_data = mol_basis_data(self.basis_name, self.symbols, load_data)
 
+        if self.unit not in ("angstrom", "bohr"):
+            raise ValueError(
+                f"The provided unit, '{unit}' is not supported. "
+                f"Please set 'unit' to 'Bohr' or 'Angstrom'."
```
Comment (soranjh) [inline pennylane/qchem/molecule.py:115]:
```suggestion
                f"The provided unit '{unit}' is not supported. "
                f"Please set 'unit' to 'bohr' or 'angstrom'."
```

File: pennylane/qchem/openfermion_obs.py [hunk 14]
```diff
@@ -997,6 +997,7 @@ def _(
 def _(
     symbols,
     coordinates,
+    unit="Bohr",
```
Comment (soranjh) [inline pennylane/qchem/openfermion_obs.py:1000]:
```suggestion
    unit="bohr",
```

File: pennylane/qchem/openfermion_obs.py [hunk 15]
```diff
@@ -1013,6 +1014,16 @@ def _(
     load_data=False,
     convert_tol=1e12,
 ):
+
+    if (coord_unit := unit.strip().lower()) not in ("angstrom", "bohr"):
+        raise ValueError(
+            f"The provided unit, '{unit}' is not supported. "
+            f"Please set 'unit' to 'Bohr' or 'Angstrom'."
```
Comment (soranjh) [inline pennylane/qchem/openfermion_obs.py:1021]:
```suggestion
            f"The provided unit '{unit}' is not supported. "
            f"Please set 'unit' to 'bohr' or 'angstrom'."
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 16]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+    """
+
+    symbols = ["H", "H"]
+    geometry_bohr = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+    geometry_ang = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.529177210903]])
+
+    hamiltonian_bohr, _ = qchem.molecular_hamiltonian(
+        symbols,
+        geometry_bohr,
+        unit="Bohr",
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:807]:
```suggestion
        unit="bohr",
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 17]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+        method=method,
+        outpath=tmpdir.strpath,
+    )
+
+    hamiltonian_ang, _ = qchem.molecular_hamiltonian(
+        symbols,
+        geometry_ang,
+        unit="Angstrom",
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:815]:
```suggestion
        unit="angstrom",
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 18]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+def test_coordinate_units_for_molecular_hamiltonian_molecule_class(method, tmpdir):
+    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units
+    based on user defined information.
+    """
+
+    symbols = ["H", "H"]
+    geometry_bohr = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+    geometry_ang = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 0.529177210903]])
+
+    molecule_bohr = qchem.Molecule(symbols, geometry_bohr, unit="Bohr")
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:839]:
```suggestion
    molecule_bohr = qchem.Molecule(symbols, geometry_bohr, unit="bohr")
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 19]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+    hamiltonian_bohr, _ = qchem.molecular_hamiltonian(
+        molecule_bohr,
+        method=method,
+        outpath=tmpdir.strpath,
+    )
+
+    molecule_ang = qchem.Molecule(symbols, geometry_ang, unit="Angstrom")
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:846]:
```suggestion
    molecule_ang = qchem.Molecule(symbols, geometry_ang, unit="angstrom")
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 20]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+    hamiltonian_ang, _ = qchem.molecular_hamiltonian(
+        molecule_ang,
+        method=method,
+        outpath=tmpdir.strpath,
+    )
+    assert qml.equal(hamiltonian_ang, hamiltonian_bohr)
+
+
+def test_unit_error_molecular_hamiltonian():
+    r"""Test that an error is raised if a wrong/not-supported units for coordinates are entered."""
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:856]:
```suggestion
    r"""Test that an error is raised if a wrong/not-supported unit for coordinates is entered."""
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 21]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+
+    symbols = ["H", "H"]
+    geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+
+    with pytest.raises(ValueError, match="The provided unit, 'degrees' is not supported."):
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:861]:
```suggestion
    with pytest.raises(ValueError, match="The provided unit 'degrees' is not supported."):
```

File: tests/qchem/test_molecule.py [hunk 22]
```diff
@@ -353,3 +353,12 @@ def test_load_basis(self, symbols, geometry, basis_name, params_ref):
         assert l == l_ref
         assert alpha == alpha_ref
         assert coeff == coeff_ref
+
+    def test_unit_error(self):
+        r"""Test that an error is raised if a wrong/not-supported units for coordinates are entered."""
```
Comment (soranjh) [inline tests/qchem/test_molecule.py:358]:
```suggestion
        r"""Test that an error is raised if a wrong/not-supported unit for coordinates is entered."""
```

File: tests/qchem/test_molecule.py [hunk 23]
```diff
@@ -353,3 +353,12 @@ def test_load_basis(self, symbols, geometry, basis_name, params_ref):
...
+
+        symbols = ["H", "H"]
+        geometry = np.array([[0.0, 0.0, 0.0], [0.0, 0.0, 1.0]])
+
+        with pytest.raises(ValueError, match="The provided unit, 'degrees' is not supported."):
```
Comment (soranjh) [inline tests/qchem/test_molecule.py:363]:
```suggestion
        with pytest.raises(ValueError, match="The provided unit 'degrees' is not supported."):
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 24]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+    [
+        "pyscf",
+        "dhf",
+        "openfermion",
+    ],
+)
+def test_coordinate_units_for_molecular_hamiltonian(method, tmpdir):
+    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units
+    based on user defined information.
+    """
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:798]:
```suggestion
def test_coordinate_units_for_molecular_hamiltonian(method, tmpdir):
    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units.
    """
```

File: tests/qchem/of_tests/test_molecular_hamiltonian.py [hunk 25]
```diff
@@ -782,3 +782,81 @@ def test_error_raised_for_missing_molecule_information():
...
+    ("method"),
+    [
+        "pyscf",
+        "dhf",
+        "openfermion",
+    ],
+)
+def test_coordinate_units_for_molecular_hamiltonian_molecule_class(method, tmpdir):
+    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units
+    based on user defined information.
```
Comment (soranjh) [inline tests/qchem/of_tests/test_molecular_hamiltonian.py:832]:
```suggestion
def test_coordinate_units_for_molecular_hamiltonian_molecule_class(method, tmpdir):
    r"""Test that molecular_hamiltonian generates the Hamiltonian for both Bohr and Angstrom units.
```

**Review by soranjh — APPROVED** [review]

**Comment by ddhawan11** [comment]
[sc-62838]

File: doc/releases/changelog-dev.md [hunk 26]
```diff
@@ -98,6 +103,7 @@
 * ``qml.QutritDepolarizingChannel`` has been added, allowing for depolarizing noise to be simulated on the `default.qutrit.mixed` device.
   [(#5502)](https://github.com/PennyLaneAI/pennylane/pull/5502)
 
+
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:106]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 27]
```diff
@@ -71,11 +71,16 @@
...
+* `qml.qchem.Molecule` and qchem functions that take Molecule object as an argument now work with coordinates
+  provided in units Angstrom along with Bohr.
+  [(#5694)](https://github.com/PennyLaneAI/pennylane/pull/5694)
+
 * The `dtype` for `eigvals` of `X`, `Y`, `Z` and `Hadamard` is changed from `int` to `float`, making them 
   consistent with the other observables. The `dtype` of the returned values when sampling these observables 
   (e.g. `qml.sample(X(0))`) is also changed to `float`. 
   [(#5607)](https://github.com/PennyLaneAI/pennylane/pull/5607)
 
+
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:83]:
```suggestion
```

File: doc/releases/changelog-dev.md [hunk 28]
```diff
@@ -71,11 +71,16 @@
   `qml.devices.Device`, which follows the new device API.
   [(#5581)](https://github.com/PennyLaneAI/pennylane/pull/5581)
 
+* `qml.qchem.Molecule` and qchem functions that take Molecule object as an argument now work with coordinates
```
Comment (trbromley) [inline doc/releases/changelog-dev.md:74]:
```suggestion
* `qml.qchem.Molecule` and qchem functions that take `Molecule` object as an argument now work with coordinates
```

**Pull request merged**